src/System.Memory/src/System/SpanSortHelpers.Keys.Comparison.cs

```diff
@@ -0,0 +1,329 @@
+            //       `int middle = lo + ((hi - lo) >> 1);`
+            int middle = (int)(((uint)hi + (uint)lo) >> 1);
+
+            // Sort lo, mid and hi appropriately, then pick mid as the pivot.
+            ref TKey keysAtLo = ref Unsafe.Add(ref keys, lo);
+            ref TKey keysAtMiddle = ref Unsafe.Add(ref keys, middle);
+            ref TKey keysAtHi = ref Unsafe.Add(ref keys, hi);
+            Sort3(ref keysAtLo, ref keysAtMiddle, ref keysAtHi, comparison);
+
+            TKey pivot = keysAtMiddle;
+
+            int left = lo;
+            int right = hi - 1;
+            // We already partitioned lo and hi and put the pivot in hi - 1.  
+            // And we pre-increment & decrement below.
+            Swap(ref keysAtMiddle, ref Unsafe.Add(ref keys, right));
+
+            while (left < right)
+            {
+                // TODO: Would be good to be able to update local ref here
+
+                while (left < (hi - 1) && comparison(Unsafe.Add(ref keys, ++left), pivot) < 0) ;
+                // Check if bad comparable/comparison
+                if (left == (hi - 1) && comparison(Unsafe.Add(ref keys, left), pivot) < 0)
+                    ThrowHelper.ThrowArgumentException_BadComparer(comparison);
+
+                while (right > lo && comparison(pivot, Unsafe.Add(ref keys, --right)) < 0) ;
+                // Check if bad comparable/comparison
+                if (right == lo && comparison(pivot, Unsafe.Add(ref keys, right)) < 0)
+                    ThrowHelper.ThrowArgumentException_BadComparer(comparison);
+
+                if (left >= right)
+                    break;
+
+                Swap(ref keys, left, right);
+            }
+            // Put pivot in the right location.
+            right = hi - 1;
+            if (left != right)
+            {
+                Swap(ref keys, left, right);
+            }
+            return left;
+        }
+
+        private static void HeapSort<TKey>(
+            ref TKey keys, int lo, int hi,
+            Comparison<TKey> comparison)
+            
+        {
+            Debug.Assert(comparison != null);
+            Debug.Assert(lo >= 0);
+            Debug.Assert(hi > lo);
+
+            int n = hi - lo + 1;
+            for (int i = n / 2; i >= 1; --i)
+            {
+                DownHeap(ref keys, i, n, lo, comparison);
+            }
+            for (int i = n; i > 1; --i)
+            {
+                Swap(ref keys, lo, lo + i - 1);
+                DownHeap(ref keys, 1, i - 1, lo, comparison);
+            }
+        }
+
+        private static void DownHeap<TKey>(
+            ref TKey keys, int i, int n, int lo,
+            Comparison<TKey> comparison)
+            
+        {
+            Debug.Assert(comparison != null);
+            Debug.Assert(lo >= 0);
+
+            //TKey d = keys[lo + i - 1];
+            ref TKey keysAtLo = ref Unsafe.Add(ref keys, lo);
+            ref TKey keysAtLoMinus1 = ref Unsafe.Add(ref keysAtLo, -1); // No Subtract??
```

Comment:
We do have Subtract - https://github.com/dotnet/corefx/blob/master/src/System.Runtime.CompilerServices.Unsafe/ref/System.Runtime.CompilerServices.Unsafe.cs#L40
```C#
 public static ref T Subtract<T>(ref T source, int elementOffset)
```

Reply:
This isn't available in corefx as of the state of this PR, we should revise once merge/or PR started over from latest corefx.
```csharp
[CLSCompliant(false)]
    public static class Unsafe
    {
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static ref T Add<T>(ref T source, int elementOffset);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static ref T Add<T>(ref T source, IntPtr elementOffset);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void* Add<T>(void* source, int elementOffset);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static ref T AddByteOffset<T>(ref T source, IntPtr byteOffset);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static bool AreSame<T>(ref T left, ref T right);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static T As<T>(object value) where T : class;
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static ref TTo As<TFrom, TTo>(ref TFrom source);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void* AsPointer<T>(ref T value);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static ref T AsRef<T>(void* source);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static IntPtr ByteOffset<T>(ref T origin, ref T target);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void InitBlockUnaligned(ref byte startAddress, byte value, uint byteCount);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static T Read<T>(void* source);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static T Read<T>(ref byte source);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static T ReadUnaligned<T>(ref byte source);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static T ReadUnaligned<T>(void* source);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static int SizeOf<T>();
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void Write<T>(void* destination, T value);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void Write<T>(ref byte destination, T value);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void WriteUnaligned<T>(void* destination, T value);
        [IntrinsicAttribute]
        [NonVersionableAttribute]
        public static void WriteUnaligned<T>(ref byte destination, T value);
    }
```